v1/topdown/regex.go

```diff
@@ -124,42 +129,48 @@ func getRegexp(bctx BuiltinContext, pat string) (*regexp.Regexp, error) {
 		if err != nil {
 			return nil, err
 		}
-		bctx.InterQueryBuiltinValueCache.Insert(ast.String(pat), re)
+		bctx.InterQueryBuiltinValueCache.Insert(key, re)
 		return re, nil
 	}
 
-	regexpCacheLock.Lock()
-	defer regexpCacheLock.Unlock()
+	regexpCacheLock.RLock()
 	re, ok := regexpCache[pat]
+	numCached := len(regexpCache)
+	regexpCacheLock.RUnlock()
 	if !ok {
 		var err error
 		re, err = regexp.Compile(pat)
 		if err != nil {
 			return nil, err
 		}
-		if len(regexpCache) >= regexCacheMaxSize {
+
+		regexpCacheLock.Lock()
+		if numCached >= regexCacheMaxSize {
 			// Delete a (semi-)random key to make room for the new one.
 			for k := range regexpCache {
 				delete(regexpCache, k)
 				break
 			}
 		}
 		regexpCache[pat] = re
+		regexpCacheLock.Unlock()
 	}
 	return re, nil
 }
 
 func getRegexpTemplate(pat string, delimStart, delimEnd byte) (*regexp.Regexp, error) {
-	regexpCacheLock.Lock()
-	defer regexpCacheLock.Unlock()
+	regexpCacheLock.RLock()
 	re, ok := regexpCache[pat]
+	regexpCacheLock.RUnlock()
 	if !ok {
 		var err error
 		re, err = compileRegexTemplate(pat, delimStart, delimEnd)
 		if err != nil {
 			return nil, err
 		}
+		regexpCacheLock.Lock()
 		regexpCache[pat] = re
+		regexpCacheLock.Unlock()
 	}
```

Comment:
If we RUnlock right after the `re, ok := ...`, wouldn't it be possible that two concurrent writers end up updating the same cache key? Not a data race, since the write is `Lock/Unlock`-protected, but a possibility to have unneeded work done. 💭

Reply:
Yeah, fair point. It'll be a temporary maybe-issue though, and once stored, possibly faster, as less is done while holding a lock. So I guess this could (ironically) be marginally worse for an integration like `regal lint`, where we race to do a million things in parallel and then exit.. while marginally better for something long-running, like `opa run --server`?

Reply:
Hard to tell? But yeah, the cache hit path is safe, as it'll never go into `if !ok`

Reply:
Hmm no, that didn't work. There doesn't seem to be a way to get the `len` of a `sync.Map`? Which makes it difficult to use with a max items cap like we currently do. Or do you see any smart way around this that I don't?